File: C++/eval.cpp
#include<iostream>
#include<stack>
#include<string>

class Solution {
  public:
  int precedence(char op) {
        if (op == '+' || op == '-')
            return 1;
        else if (op == '*')
            return 2;
        return 0;
    }

    int compute(char op, int a, int b) {
        if (op == '+')
            return a + b;
        else if (op == '-')
            return a - b;
        else if (op == '*')
            return a * b;
        return 0; 
    }

    int solve(std::string s) {
        std::stack<int> operands;
        std::stack<char> operators;

        for (int i = 0; i < s.length(); i++) {
            if (s[i] == ' ')
                continue;
            else if (isdigit(s[i])) {
                int num = 0;
                while (i < s.length() && isdigit(s[i])) {
                    num = num * 10 + (s[i] - '0');
                    i++;
                }
                operands.push(num);
                i--;
            } 
            else if (s[i] == '(') {
                operators.push(s[i]);
            } 
            else if (s[i] == ')') {
                while (operators.top() != '(') {
                    char op = operators.top();
                    operators.pop();
                    int b = operands.top();
                    operands.pop();
                    int a = operands.top();
                    operands.pop();
                    operands.push(compute(op, a, b));
                }
                operators.pop(); 
            } 
            else {
                while (!operators.empty() && precedence(operators.top()) >= precedence(s[i])) {
                    char op = operators.top();
                    operators.pop();
                    int b = operands.top();
                    operands.pop();
                    int a = operands.top();
                    operands.pop();
                    operands.push(compute(op, a, b));
                }
                operators.push(s[i]);
            }
        }

        while (!operators.empty()) {
            char op = operators.top();
            operators.pop();
            int b = operands.top();
            operands.pop();
            int a = operands.top();
            operands.pop();
            operands.push(compute(op, a, b));
        }

        return operands.top();
    }
};

int main(){
    std::string s="(2*(3-4))*5";
    Solution solution;
    std::cout<<solution.solve(s)<<std::endl;
    return 0;
}

File: python/eval.py
class Solution:
    def precedence(self, op: str) -> int:
        if op in {"+", "-"}:
            return 1
        if op == "*":
            return 2
        return 0

    def compute(self, op: str, a: int, b: int) -> int:
        if op == "+":
            return a + b
        elif op == "-":
            return a - b
        elif op == "*":
            return a * b
        return 0

    def solve(self, s: str) -> int:
        operand = []
        operator = []

        i = 0
        while i < len(s):
            if s[i] == " ":
                i += 1
                continue
            elif s[i].isdigit():
                num = 0
                while i < len(s) and s[i].isdigit():
                    num = num * 10 + int(s[i])
                    i += 1
                operand.append(num)
                i -= 1
            elif s[i] == "(":
                operator.append(s[i])
            elif s[i] == ")":
                while operator[-1] != "(":
                    op = operator[-1]
                    operator.pop()
                    b = operand[-1]
                    operand.pop()
                    a = operand[-1]
                    operand.pop()
                    operand.append(self.compute(op, a, b))
                operator.pop()
            else:
                while operator and self.precedence(operator[-1]) >= self.precedence(s[i]):
                    op = operator[-1]
                    operator.pop()
                    b = operand[-1]
                    operand.pop()
                    a = operand[-1]
                    operand.pop()
                    operand.append(self.compute(op, a, b))
                operator.append(s[i])
            i += 1

        while operator:
            op = operator[-1]
            operator.pop()
            b = operand[-1]
            operand.pop()
            a = operand[-1]
            operand.pop()
            operand.append(self.compute(op, a, b))

        return operand[-1]

s1="(2*(3-4))*5"
s2="3+2*3*4-1"
s=Solution()
print(s.solve(s1))
print(s.solve(s2))